volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,437 @@
+            ),
+            requirements.VersionRequirement(
+                name="pdbutil", component=pdbutil.PDBUtility, version=(1, 0, 0)
+            ),
+            requirements.ChoiceRequirement(
+                name="source",
+                description="Where to resolve symbols.",
+                choices=["kernel", "processes"],
+                optional=False,
+            ),
+            requirements.StringRequirement(
+                name="module",
+                description='Module in which to resolve symbols. Use "ntoskrnl.exe" to resolve in the base kernel executable.',
+                optional=False,
+            ),
+            requirements.StringRequirement(
+                name="symbol",
+                description="Symbol name to resolve",
+                optional=False,
+            ),
+        ]
+
+    @staticmethod
+    def _get_pefile_obj(
+        context, pe_table_name: str, layer_name: str, base_address: int
+    ) -> Optional[pefile.PE]:
+        """
+        Attempts to pefile object from the bytes of the PE file
+
+        Args:
+            pe_table_name: name of the pe types table
+            layer_name: name of the process layer
+            base_address: base address of the module
+
+        Returns:
+            the constructed pefile object
+        """
+        pe_data = io.BytesIO()
+
+        try:
+            dos_header = context.object(
+                pe_table_name + constants.BANG + "_IMAGE_DOS_HEADER",
+                offset=base_address,
+                layer_name=layer_name,
+            )
+
+            for offset, data in dos_header.reconstruct():
+                pe_data.seek(offset)
+                pe_data.write(data)
+
+            pe_ret = pefile.PE(data=pe_data.getvalue(), fast_load=True)
+
+        except exceptions.InvalidAddressException:
+            pe_ret = None
+
+        return pe_ret
+
+    @staticmethod
+    def _find_through_exports(
+        context,
+        config_path: str,
+        mod_name: str,
+        vads: Tuple[int, int, int],
+        wanted_symbols: Dict[str, List[str]],
+        found_symbols: Dict[str, Tuple[str, int]],
+    ):
+        """
+        Attempts to resolve the symbols in `wanted_symbols` through export table analysis
+        """
+        pe_table_name = intermed.IntermediateSymbolTable.create(
+            context, config_path, "windows", "pe", class_types=pe.class_types
+        )
+
+        # for each process layer and VAD, construct a PE and examine the export table
+        for proc_layer_name, vad_start, _ in vads:
+            # we need a valid PE with an export table
+            pe_module = PESymbols._get_pefile_obj(
+                context, pe_table_name, proc_layer_name, vad_start
+            )
+            if not pe_module:
+                continue
+
+            pe_module.parse_data_directories(
+                directories=[pefile.DIRECTORY_ENTRY["IMAGE_DIRECTORY_ENTRY_EXPORT"]]
```

Comment:
Can this throw a KeyError, or does the `hasattr` check below catch all the possible issues?

Reply:
the hasattr is how the API says to do / is in the examples.

Reply:
Ok